Windows "app" support

jasonrclark opened the pull request:
As discussed over in the [Shoes packaging plan](https://github.com/shoes/shoes4/issues/1415), our current aim is to get a "good enough" sort of distributable packaged up for all 3 OS's. This is the rough beginnings of support for Windows!

The moving pieces here are very similar to `Furoshiki::JarApp` with regards to building out a tmp dir based on a downloaded template zip file, then plugging the right pieces in. Plenty to be revised, consolidated, etc. here between the two flavors.

## Where's My Java?
There's one big difference between this and the Mac template, though, and that's the handling of the JRE.

The Mac template embeds a full JRE (roughly ~65MB) so that the thing can be guaranteed to run standalone without any intervention.

If you drag down the Windows template zip, it looks like this instead, clocking in at less than 2MB:

```
Archive:  /Users/jason/.furoshiki/cache/windows-app-template-0.0.1.zip
  Length     Date   Time    Name
 --------    ----   ----    ----
        0  04-02-17 21:41   windows-app-template/
      574  04-02-17 21:41   windows-app-template/app.bat
        0  04-02-17 21:17   windows-app-template/helpers/
    69120  04-02-17 21:17   windows-app-template/helpers/bzip2.dll
  1177600  04-02-17 21:17   windows-app-template/helpers/libeay32.dll
  1008128  04-02-17 21:17   windows-app-template/helpers/libiconv2.dll
   103424  04-02-17 21:17   windows-app-template/helpers/libintl3.dll
   232960  04-02-17 21:17   windows-app-template/helpers/libssl32.dll
   164864  04-02-17 21:17   windows-app-template/helpers/unzip.exe
   162816  04-02-17 21:17   windows-app-template/helpers/unzip32.dll
   449024  04-02-17 21:17   windows-app-template/helpers/wget.exe
```

What's up with that? Am I just assuming that the user will have a JRE available? Not at all!

[`app.bat`](https://github.com/jasonrclark/windows-app-templates/blob/master/base/app.bat) looks around to see if it can run `java`, and if it can't, it downloads a suitable JRE from our [Github releases](https://github.com/jasonrclark/windows-app-templates/releases). It tucks this JRE in a well-known spot (`~/.shoes/jdk`) so subsequent Shoes app runs can find and use it.

This approach allows folks that already have a JVM to just use it, while others will get something workable without having to make every package massive. Thoughts?

## Smaller Notes

* Startup is currently just a `.bat` but before 4.0 I want to replace that with a small native Windows binary that'll do a better job than Windows batch coding :shudders:
* Template zip is hosted as a Github release on https://github.com/jasonrclark/windows-app-templates. Currently still using the explicit version binding from `furoshiki` to the template it uses
* Simplified things in the template by assuming we'll copy the jar to `app.jar`. Now we only need to rename the start executable (currently `app.bat`) but the executable can look to a standard location for the actual jar to run.

ccoupe commented:
>  I want to replace that with a small native Windows binary that'll do a better job than Windows batch coding :shudders:

Shoes 3.3 has a C native launcher in platforms/msw that  you can modify but it bleeds assumptions up the food chain that Shoes 4 may not like.

jasonrclark commented:
Appreciate the pointer @ccoupe. Will definitely look there as a starting place when I get to building out that binary. While it will almost certainly take a different form eventually, you've got a ton of hard-won experience in writing native Windows code there that I'd be foolish to ignore!

ccoupe commented:
Be aware that Windows 10 really, really wants code signed exe's from certificate authorities that  MSFT likes. Big Dollars for one of those, per year, every year. If your user's create an  launcher  exe (as users would do ) then they need a code certificate or you vouch for their code. Rock. Hard Spot.

pjfitzgibbons commented:
Some of these might be affordable.  It appears that depending upon who you certify with, *someone* needs to personally identify to the certificate auth as the "owner" of the certificate, including their national ID and permanent residence address.

Here's what I see as referenced from https://msdn.microsoft.com/en-us/windows/hardware/drivers/dashboard/get-a-code-signing-certificate
Plus google search "Microsoft code signing certificate open source projects"

Certum - Open Source Certificate
https://www.certum.eu/certum/cert,offer_en_open_source_cs.xml
Can't tell from the details if this is a .exe signing certificate.
€86.00 gross

The SSLStore - https://www.thesslstore.com/promoads/cheap-code-signing.aspx
COMODO branded
$85/yr


Symantec (https://products.websecurity.symantec.com/orders/enrollment/winqualOrder.do)
No price given - "CSR generation failed. CSR required to continue"
* shrug *

Certum - https://en.sklep.certum.pl/data-safety/code-signing-certificates/standard-code-signing.html
€154.12 Incl. Tax / Yr

Entrust - https://www.entrust.com/code-signing-certificates/
$524.25 / yr

Globalsign - https://www.globalsign.com/en/code-signing-certificate/
$289/yr

Digicert - https://www.digicert.com/order/order-1.php
$224/yr

pjfitzgibbons commented:
Honestly, at the < $200 range, if at least one other would go halfsies with me on it I would happily fund the windows signing cert.

ccoupe commented:
If your launcher attempts to install an exe (as Shoes 3.3 does) and call it to to eval a script/shy (as Shoes 3.3 does) then Win 10 thinks it's virus. - Technically, it could be a virus/trojan/bad-actor. For Shoes4, installing java,  if needed, will need a certificate - odds are java is signed. Invoking that java to install JRuby - might work on win10 until the overlords at MSFT catch that pattern in their AV engine and want a cert from jRuby. Invoking the freshly installed Jruby to install Shoe4 and then call that to run an unknown ruby script from jrandom_user?  

As I said, there are "bubble up" issues design changes that follow.  Did I  mention the horrible code singing dance which can only be done on Windows? Shoes 3.3 can package a users app (shy)  like HacketyHack did and we punt the code signing problem to the user and installer customization (NSIS or QtIFW).  His app, not ours.

jasonrclark commented:
@pjfitzgibbons Thanks for the offer, and if it comes down to that I'd happily fund things as well.

Haven't done any further research, but appreciate all the hard-won wisdom you're providing @ccoupe!

Hoping to get some time tonight to finally revise this and get it to a happy-enough state before EOW. Shoes on!

jasonrclark commented:
Moved over the template file repo to `shoes` org, updated this to match and I think this is ready for go-time. Will merge over the weekend unless someone sees something that needs updating.

jasonrclark commented:
Back from RailsConf and ready to start moving again. I'm going to merge this and then just roll forward with any changes needed.
